cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/connector/Connector.java

```diff
@@ -38,26 +41,33 @@ default void configure(PluginConfigurer configurer) {
 
   /**
    * Test if the connector is able to connect to the resource
+   *
+   * @param collector a collector used to collect the failures
+   * @throws ValidationException if the connector is not able to connect to the source
    */
-  void test() throws Exception;
+  void test(FailureCollector collector) throws ValidationException;
 
   /**
    * Explore the resource on the given request. The explore expects a path to represent the hierarchy of the
    * resource. The path is expected to be separated by '/'.
    * For example, for a file based connector, the path will just be the file/directory path
    * for a database connector, the path can be {database}/{table}
+   *
+   * @param request the browse request
    */
-  ExploreDetail explore(ConnectRequest request) throws Exception;
+  BrowseDetail browse(BrowseRequest request) throws IOException;
 
   /**
-   * Generate the properties based on the connector config, the path and and sampling properties
-   * @param samplingProperties the sampling properties of the entity in the path
-   * @return the map which contains all the properties which can be used by a source or a sink
+   * Parse the given path, generates the properties associated with the path.
+   * For example, for bigquery, this can be a map containing "datasets": {dataset}, "table": {table}
+   *
+   * @param path the path of the entitiy
+   * @return the map which contains all the properties associated with the path
    */
-  Map<String, String> generateSpec(String path, Map<String, String> samplingProperties);
+  Map<String, String> parsePath(String path);
```

Comment:
I think the old name of generateSpec is better, as it's more descriptive of what is returned rather than how it is returned. It would also be better to return an object, like a ConnectorSpec, even if all it has right now is a map of properties. This would allow us to extend the API easily without breaking compatibility.